server/sonar-process-monitor/src/main/java/org/sonar/application/cluster/AppStateClusterImpl.java

```diff
@@ -0,0 +1,208 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.application.cluster;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.config.JoinConfig;
+import com.hazelcast.config.NetworkConfig;
+import com.hazelcast.core.EntryEvent;
+import com.hazelcast.core.EntryListener;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IAtomicReference;
+import com.hazelcast.core.ILock;
+import com.hazelcast.core.MapEvent;
+import com.hazelcast.core.ReplicatedMap;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import org.sonar.application.config.AppSettings;
+import org.sonar.application.AppState;
+import org.sonar.application.AppStateListener;
+import org.sonar.process.ProcessId;
+
+public class AppStateClusterImpl implements AppState, AutoCloseable {
+  static final String OPERATIONAL_PROCESSES = "operational_processes";
+  static final String LEADER = "leader";
+
+  private final List<AppStateListener> listeners = new ArrayList<>();
+  private final ReplicatedMap<ClusterProcess, Boolean> operationalProcesses;
+  private final String listenerUuid;
+
+  final HazelcastInstance hzInstance;
+
+  public AppStateClusterImpl(AppSettings appSettings) {
+    ClusterProperties clusterProperties = new ClusterProperties(appSettings);
+    clusterProperties.validate();
+
+    if (!clusterProperties.isEnabled()) {
+      throw new IllegalStateException("Cluster is not enabled on this instance");
+    }
+
+    Config hzConfig = new Config();
+    try {
+      hzConfig.setInstanceName(InetAddress.getLocalHost().getHostName());
+    } catch (UnknownHostException e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)